#!/bin/bash

# SLURM parameters
# job standard output will go to the file slurm-%j.out (where %j is the job ID)

#SBATCH --job-name="GBS variant calling - variant calling"
#SBATCH -p short
#SBATCH -t 24:00:00   # walltime limit (HH:MM:SS)
#SBATCH -N 1   # number of nodes
#SBATCH -n 72   # 8 processor core(s) per node X 2 threads per core
#SBATCH --mem=216G   # maximum memory per node
#SBATCH --mail-user=jeffrey.neyhart@usda.gov   # email address
#SBATCH --mail-type=BEGIN,END,FAIL


##
## Capture Seq Variant Calling Pipeline
##
## Step 3. merge alignment files and call variants using freebayes
##

# Set error handling options
set -e
set -u
set -o pipefail

# Load the modules
module load samtools
module load freebayes

#####################
## Set variables
#####################

# A name for the project
PROJNAME="cranberry_prior2021_gbs"

# Working directory
WD=/project/gifvl_vaccinium/cranberryGenotyping/cranberryHistoricalGBS

# Prefix of the indexed reference genome
REFPREFIX=/project/gifvl_vaccinium/cranberryGenotyping/genome_assemblies/Vaccinium_macrocarpon_Stevens_v1.fasta

# Directory to the alignment files
ALIGNDIR=$WD/alignment
# Directory to store the merged alignment file
MERGEDALIGNDIR=$WD/merged_alignment

# Directory to output variants
VARIANTDIR=$WD/variants/

# Number of threads available
NTHREADS=$SLURM_JOB_CPUS_PER_NODE


##############################
## DO NOT EDIT BELOW
##############################


## Run the pipeline

# Change working directory
cd $WD

# ## Merge BAM files
# # List the bam files
# BAMFILES=$(find $ALIGNDIR -name "*_alignment.bam")
# 
# # Merge the bam files
# # Sort on coordinates
# samtools merge -@ $SLURM_JOB_CPUS_PER_NODE -o - $BAMFILES | \
# 	samtools sort -@ $SLURM_JOB_CPUS_PER_NODE -u - | \
# 	samtools view -b -o $MERGEDALIGNDIR/${PROJNAME}_alignments_merged.bam -@ $SLURM_JOB_CPUS_PER_NODE -
# 
# # Index
# samtools index $MERGEDALIGNDIR/${PROJNAME}_alignments_merged.bam


# List alignment files
ALIGNMENTFILES=$(find $MERGEDALIGNDIR -name "*alignments_merged.bam")

## Run variant calling 
# Ouput file
OUTPUT=$VARIANTDIR/${PROJNAME}_variants.vcf

# Use parallelization

### IMPORTANT ###
## DO NOT USE THE -t OPTION WITH PARALLELIZATION; IT WILL ATTEMPT TO SEARCH FOR ALL POSITIONS IN THE BED FILE
## FOR EVERY SPLIT IN THE REFERENCE THAT IS GENERATED BY fasta_generate_regions.py
#################

freebayes-parallel <(fasta_generate_regions.py $REFPREFIX 1000000) $SLURM_JOB_CPUS_PER_NODE \
	-f $REFPREFIX --use-best-n-alleles 2 --ploidy 2 $ALIGNMENTFILES > $OUTPUT

# gzip the output
gzip -f $OUTPUT
