---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# GBS - Scripts

<!-- badges: start -->
<!-- badges: end -->

The subfolder contains the scripts used to process reads generated via genotyping-by-sequencing and call SNP variants. The order and description of the scripts is below:

## Read processing and variant calling

1. `01_run_tassel5_pipeline.sh` - runs the entire TASSEL5 variant-calling pipeline.

2. `02_filter_variants_allUniqueKeys.sh` - filters the variants generated from Step 2 based on the following: MinGQ = 40, MinMeanDP = 10, MinDP = 7, and bi-allelic SNP variants only.

## Imputation and other post-processing

7. `07_run_beagle_imputation.sh` - runs BEAGLE to impute variants.

8. `08_run_plink_stats.sh` - generates some haplotype stats using PLINK.
